rs/sns/governance/src/extensions.rs

```diff
@@ -40,35 +40,246 @@ impl Display for ExtensionKind {
+                    validate: validate_deposit_operation,
+                },
+                OperationSpec::GovernanceSupported {
+                    name: "withdraw".to_string(),
+                    description: "Withdraw funds from the treasury manager.".to_string(),
+                    extension_type: ExtensionKind::TreasuryManager,
+                    validate: validate_withdraw_operation,
+                },
+            ],
+            // Future extension types would define their standard operations here
+        }
+    }
 }
 
+#[derive(Clone, Debug, Eq, PartialEq)]
+struct ExtensionVersion(u64);
+
 #[derive(Clone, Debug, Eq, PartialEq)]
 pub struct ExtensionSpec {
     pub name: String,
+    pub version: ExtensionVersion,
     pub topic: Topic,
-    pub kind: ExtensionKind,
-    pub operations: BTreeMap<String, ExtensionOperationSpec>,
+    /// The extension types this extension implements (can be multiple)
+    pub extension_types: Vec<ExtensionKind>,
+    /// Additional operations beyond the standard ones from extension_types
+    pub other_operations: BTreeMap<String, OperationSpec>,
     // TODO: Add a way to specify initialization arguments schema for the extension.
 }
 
+impl ExtensionSpec {
+    /// Validates that there are no operation name conflicts
+    pub fn validate(&self) -> Result<(), String> {
+        // This restriction may be relaxed later, but at present each extension type can only
+        // have one responsibility.
+        if self.extension_types.len() > 1 {
+            return Err("ExtensionSpec can only have one extension type at a time".to_string());
+        }
+
+        let mut seen_operations = BTreeMap::new();
+
+        // First, collect all standard operations
+        for ext_type in &self.extension_types {
+            for op in ext_type.standard_operations() {
+                let name = op.name();
+                if let Some(existing_type) = seen_operations.insert(name.to_string(), ext_type) {
+                    return Err(format!(
+                        "Operation '{}' is defined by multiple extension types: {:?} and {:?}",
+                        name, existing_type, ext_type
+                    ));
+                }
+            }
+        }
+
+        // Check that other_operations don't conflict with standard operations
+        for (name, _) in &self.other_operations {
+            if seen_operations.contains_key(name) {
+                return Err(format!(
+                    "Custom operation '{}' conflicts with a standard operation from extension type {:?}",
+                    name, seen_operations.get(name).unwrap()
+                ));
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Get all operations (standard + other) for this extension
+    /// Returns error if there are conflicts
+    pub fn all_operations(&self) -> Result<BTreeMap<String, OperationSpec>, String> {
+        self.validate()?;
```

Comment:
Add todo for validation (name conflicts) when more than 1 extension type is supported, or we get the custom functions exposed by extensions